### PR TITLE
Create signature controller package

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## Description

Create the `@metamask/signature-controller` package to contain the `SignController` from the extension so it can also be adopted in mobile.

This controller encapsulates usage of the message managers to process signing requests from the `eth_sign`, `personal_sign` and `eth_signTypedData` methods.

## Changes

- **ADDED:** `@metamask/signature-controller` package

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/signature-controller/src/SignatureController.ts [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+  AcceptRequest,
+  AddApprovalRequest,
+  RejectRequest,
+} from '@metamask/approval-controller';
+import { ORIGIN_METAMASK } from '@metamask/controller-utils';
+
+const controllerName = 'SignatureController';
+const methodNameSign = 'eth_sign';
+const methodNamePersonalSign = 'personal_sign';
+const methodNameTypedSign = 'eth_signTypedData';
```
Comment (bergarces) [inline packages/signature-controller/src/SignatureController.ts:37]: These are used as approval types, so they should probably use `ApprovalType` enum.

File: packages/signature-controller/src/SignatureController.ts [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+      ...this.state.unapprovedTypedMessages,
+    }[messageId];
+  }
+
+  private _requestApproval(
+    msgParams: AbstractMessageParamsMetamask,
+    type: string,
+  ) {
+    const id = msgParams.metamaskId as string;
+    const origin = msgParams.origin ?? ORIGIN_METAMASK;
```
Comment (bergarces) [inline packages/signature-controller/src/SignatureController.ts:638]: I'm unsure about what exact behaviour we want here, but with `??`, `ORIGIN_METAMASK` will be used only if `msgParams.origin` is `undefined`.

If we were to use `||` it'll be used if `msgParams.origin` is falsy, such as `''`.

File: packages/signature-controller/src/SignatureController.ts [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    messageId: string,
+    reason?: string,
+  ) {
+    if (reason) {
+      const message = this._getMessage(messageId);
+      this.hub.emit('cancelWithReason', { message, reason });
```
Comment (bergarces) [inline packages/signature-controller/src/SignatureController.ts:535]: Is the plan to add an event listener for this event to the extension and mobile? I cannot find an existing one.

File: packages/signature-controller/src/SignatureController.ts (hunk 1 above)
Comment (vinistevam) [reply packages/signature-controller/src/SignatureController.ts:37]: :heavy_plus_sign:

File: packages/signature-controller/src/SignatureController.ts [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+  unapprovedTypedMessagesCount: 0,
+});
+
+type CoreMessage = AbstractMessage & {
+  messageParams: AbstractMessageParams;
+};
+
+type StateMessage = Required<AbstractMessage> & {
+  msgParams: Required<AbstractMessageParams>;
+  securityProviderResponse: any;
```
Comment (vinistevam) [inline packages/signature-controller/src/SignatureController.ts:63]: Didn't you need to Omit `securityProviderResponse` from AbstractMessage? That would save me a lot of time.

Comment (bergarces) [reply packages/signature-controller/src/SignatureController.ts:63]: Now that I think about it, the `Omit` suggestion was good for the encryption public key message because there isn't one, but in this case, that response is sent to the UI, so it should be there.

When we do the ticket about improving MessageManager types we can tackle all these details better.

File: packages/signature-controller/src/SignatureController.ts (hunk 1 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: Done.

File: packages/signature-controller/src/SignatureController.ts (hunk 2 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: I was trying to be more explicit to show we're handling an undefined origin, but you're right, we should handle the empty string case too. Done.

File: packages/signature-controller/src/SignatureController.ts (hunk 3 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts:531]: Exactly. This was the last thing I did in the extension, so was still testing at the time.

File: packages/signature-controller/src/SignatureController.ts (hunk 4 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: Forgot about this, I've removed it completely for clarity and fixed the type in `AbstractMessageManager` to remove the type error.

I've also removed the use of the draft state which was ultimately causing the infinite error.

**Review by vinistevam — APPROVED** [review]
Well done! First new controller :1st_place_medal:

File: packages/signature-controller/LICENSE [hunk 5]
```diff
@@ -0,0 +1,20 @@
+MIT License
+
+Copyright (c) 2018 MetaMask
```
Comment (Mrtenz) [inline packages/signature-controller/LICENSE:3]:
```suggestion
Copyright (c) 2023 MetaMask
```

File: packages/signature-controller/src/SignatureController.ts [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+
+    this._keyringController = keyringController;
+    this._isEthSignEnabled = isEthSignEnabled;
+    this._getState = getState;
```
Comment (mcmire) [inline packages/signature-controller/src/SignatureController.ts:165]: Since the controller has state, I wonder if consumers could get confused into thinking that this has to do with the controller. Whose state is this? Should this option be renamed to something clearer?

File: packages/signature-controller/src/SignatureController.test.ts [hunk 7]
```diff
@@ -0,0 +1,740 @@
...
+  TypedMessageManager: jest.fn(),
+}));
+
+jest.mock('@metamask/controller-utils', () => {
+  const actual = jest.requireActual('@metamask/controller-utils');
+  return { ...actual, detectSIWE: jest.fn() };
+});
+
+// eslint-disable-next-line jest/prefer-spy-on
+console.info = jest.fn();
```
Comment (mcmire) [inline packages/signature-controller/src/SignatureController.test.ts:27]: This will change the behavior of `console.info` across all tests. Is that what we want? I think we ought to follow what the ESLint rule says and use `spyOn`.

File: packages/message-manager/src/AbstractMessageManager.ts [hunk 8]
```diff
@@ -32,7 +32,7 @@ export interface AbstractMessage {
   status: string;
   type: string;
   rawSig?: string;
-  securityProviderResponse?: Map<string, Json>;
+  securityProviderResponse?: Record<string, Json>;
```
Comment (mcmire) [inline packages/message-manager/src/AbstractMessageManager.ts:35]: This seems like a breaking change. Maybe it would be better to extract this to another PR so that it's more noticeable?

**Review by mcmire — COMMENTED** [review]
Hey @matthewwalsh0, I did a shallow pass over this PR. I see you figured out how to create a new package so well done! Hopefully you didn't have too much trouble doing so, but if there's any pain points you found or any part of the setup you were wondering about, feel free to let us know.

In any case, I noticed a couple of things that I think we ought to address — the change to AbstractMessage as well as the version in `package.json` — but the other comments I have are more minor. I did have some suggestions on writing tests. They aren't necessarily blockers but it might make things more understandable and keep us from committing some TypeScript sins.

**Comment by matthewwalsh0** [comment]
@SocketSecurity ignore @cspotcode/source-map-support@0.8.1

File: packages/message-manager/src/AbstractMessageManager.ts (hunk 8 above)
Comment (matthewwalsh0) [reply packages/message-manager/src/AbstractMessageManager.ts:35]: I have preview PRs ready in both the extension and client so can confirm this has been tested, this is purely to simplify the typing of the new controller as `BaseControllerV2` doesn't support `Map` in controller state.

File: packages/signature-controller/src/SignatureController.test.ts (hunk 7 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts]: Done.

File: packages/signature-controller/src/SignatureController.ts (hunk 6 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: Done, renamed to `getAllState`.

File: packages/signature-controller/src/SignatureController.ts [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+   * Clears all unapproved messages from memory.
+   */
+  clearUnapproved() {
+    this._clearUnapproved(this._messageManager);
+    this._clearUnapproved(this._personalMessageManager);
+    this._clearUnapproved(this._typedMessageManager);
+  }
+
+  private _rejectUnapproved<
+    M extends AbstractMessage,
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts:452]: The controller contains three different types of message manager, hence the need for generic methods and these templates.

File: packages/signature-controller/src/SignatureController.ts [hunk 10]
```diff
@@ -0,0 +1,678 @@
...
+    messsageParams: TypedMessageParams,
+    options: { version: string | undefined },
+  ) => Promise<string>;
+}
+
+export type SignatureControllerOptions = {
+  messenger: SignatureControllerMessenger;
+  keyringController: KeyringController;
+  isEthSignEnabled: () => boolean;
+  getState: () => any;
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: Done.

File: packages/signature-controller/src/SignatureController.ts [hunk 11]
```diff
@@ -0,0 +1,678 @@
...
+   *
+   * @param msgParams - The params passed to eth_sign.
+   * @param [req] - The original request, containing the origin.
+   * @returns Promise resolving to the raw data of the signature request.
+   */
+  async newUnsignedMessage(
+    msgParams: MessageParams,
+    req: OriginalRequest,
+  ): Promise<string> {
+    // eslint-disable-next-line camelcase
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: This wasn't required, but copied from the extension, removed.

File: packages/signature-controller/src/SignatureController.ts [hunk 12]
```diff
@@ -0,0 +1,678 @@
...
+ * Controller for creating signing requests requiring user approval.
+ */
+export class SignatureController extends BaseControllerV2<
+  typeof controllerName,
+  SignatureControllerState,
+  SignatureControllerMessenger
+> {
+  hub: EventEmitter;
+
+  private _keyringController: KeyringController;
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: Done.

File: packages/signature-controller/src/SignatureController.test.ts [hunk 13]
```diff
@@ -0,0 +1,740 @@
...
+  });
+
+  describe('message manager events', () => {
+    it.each([
+      ['message manager', messageManagerMock],
+      ['personal message manager', personalMessageManagerMock],
+      ['typed message manager', typedMessageManagerMock],
+    ])('bubbles update badge event from %s', (_, messageManager) => {
+      const mockListener = jest.fn();
+      const mockHub = messageManager.hub.on as jest.Mock;
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts:621]: Not really, as it's a nested property, we need to clarify it's mocked. I could use `mocked` but that feels like overkill as it's really just another type of casting.

File: packages/signature-controller/src/SignatureController.test.ts [hunk 14]
```diff
@@ -0,0 +1,740 @@
...
+
+  describe('rejectUnapproved', () => {
+    beforeEach(() => {
+      const messages = {
+        [messageIdMock]: messageMock,
+        [messageIdMock2]: messageMock,
+      };
+
+      messageManagerMock.getUnapprovedMessages.mockReturnValueOnce(
+        messages as any,
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts:208]: For the sake of less code, I'm creating a single mock object above using abstract messages, hence the need to cast.

File: packages/signature-controller/src/SignatureController.test.ts [hunk 15]
```diff
@@ -0,0 +1,740 @@
...
+    personalMessageManagerConstructorMock.mockReturnValue(
+      personalMessageManagerMock,
+    );
+
+    typedMessageManagerConstructorMock.mockReturnValue(typedMessageManagerMock);
+
+    isEthSignEnabledMock.mockReturnValue(true);
+
+    signatureController = new SignatureController({
+      messenger: messengerMock as any,
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts]: Done, none of these were needed.

File: packages/signature-controller/src/SignatureController.test.ts [hunk 16]
```diff
@@ -0,0 +1,740 @@
...
+    approveMessage: jest.fn(),
+    setMessageStatusSigned: jest.fn(),
+    setMessageStatusErrored: jest.fn(),
+    rejectMessage: jest.fn(),
+    subscribe: jest.fn(),
+    update: jest.fn(),
+    hub: {
+      on: jest.fn(),
+    },
+  }) as jest.Mocked<T>;
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts:92]: Unfortunately, it's an awkward case where I can't use real instances as I need to mock the constructors to get access to the message manager instances in order to assert against them.

I can't use a fake implementation either as I'm using multiple different concrete instances in the controller which need to pass `instanceof` checks, hence the use of `Object.create` to set the prototype.

File: packages/signature-controller/package.json [hunk 17]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@metamask/signature-controller",
+  "version": "1.0.0",
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/package.json]: Done.

**Review by bergarces — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]
